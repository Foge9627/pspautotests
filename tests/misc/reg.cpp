#include <common.h>
#include <stdlib.h>
#include <malloc.h>
#include <pspthreadman.h>
#include <pspreg.h>
#include <string>
#include <vector>

#define SCE_REG_ERROR_CATEGORY_EMPTY_MAYBE 0x80082712

void DumpCategory(REGHANDLE regHandle, const std::string &path, const std::string &concat_name, const std::string &name) {
    REGHANDLE category = 0xcccccccc;

    int retval = sceRegOpenCategory(regHandle, path.c_str(), 2, &category);
    if (retval == SCE_REG_ERROR_CATEGORY_EMPTY_MAYBE) {
        schedf("// %s was not accessible (returned %08x)\n", path.c_str(), retval);
        schedf("static const KeyValue %s[1] = { \"\", ValueType::FAIL, \"\", (int)0x%08x };\n\n", concat_name.c_str(), retval);
        return;
    }

    if (retval < 0) {
        schedf("%08x = sceRegOpenCategory(%s)\n", retval, path.c_str());
        return;
    }

    int numKeys;
    int result = sceRegGetKeysNum(category, &numKeys);
    if (result < 0) {
        schedf("%08x = sceRegGetKeysNum(%s) -> %d\n", result, numKeys);
        return;
    }
    // schedf("%08x = sceRegGetKeysNum(fontCategory) -> %d\n", result, numKeys);

    char *keyData = (char *)malloc(numKeys * 27);  // ??? 27 bytes per key?
    result = sceRegGetKeys(category, keyData, numKeys);

    // First, dump all subcategories.
    for (int i = 0; i < numKeys; i++) {
        char *keyName = keyData + i * 27;
        unsigned int type;
        SceSize size = 0xcccccccc;
        unsigned char data[1024];
        REGHANDLE keyHandle;
        if (!sceRegGetKeyInfo(category, keyName, &keyHandle, &type, &size)) {
            if (type == REG_TYPE_DIR) {
                DumpCategory(regHandle, path + "/" + keyName, concat_name + "_" + keyName, keyName);
            }
        }
    }

    // Then, actually list the contents of this category, in the struct form.
    schedf("// Dump of %s\n", path.c_str());
    schedf("static const KeyValue %s[] = {\n", concat_name.c_str());
    for (int i = 0; i < numKeys; i++) {
        char *keyName = keyData + i * 27;
        unsigned int type;
        SceSize size = 0xcccccccc;
        unsigned char data[1024];
        REGHANDLE keyHandle;
        if (!sceRegGetKeyInfo(category, keyName, &keyHandle, &type, &size)) {
            if (type == REG_TYPE_DIR) {
                std::string subdirName = concat_name + "_" + keyName;
                schedf("\t{ \"%s\", ValueType::DIR, \"\", ARRAY_SIZE(%s), %s },\n", keyName, subdirName.c_str(), subdirName.c_str());
            } else {
                if (!sceRegGetKeyValue(category, keyHandle, data, size)) {
                    switch(type) {
                    case REG_TYPE_INT:
                    {
                        int ivalue = *((int*) data);
                        schedf("\t{ \"%s\", ValueType::INT, \"\", (int)0x%08x },\n", keyName, ivalue);
                        break;
                    }
                    case REG_TYPE_STR:
                        schedf("\t{ \"%s\", ValueType::STR, \"%s\" },\n", keyName, (char *)data);
                        break;
                    case REG_TYPE_BIN:
                        // Not yet supported
                        schedf("\t // Skipping %s (binary)\n", keyName);
                        break;
                    };
                }
            }
        }
    }

    schedf("};\n\n");

    free(keyData);
}

void DumpRegistry(REGHANDLE regHandle) {
    schedf("\n\n// Dump of the PSP registry using tests/misc/reg.prx in pspautotests\n\n");

    static const char *topLevel[] = {
        "DATA",
        "SYSPROFILE",
        // probably more...
    };

    for (int i = 0; i < ARRAY_SIZE(topLevel); i++) {
        std::string concat = "tree_" + std::string(topLevel[i]);
        for (int j = 0; j < concat.size(); j++) {
            if (concat[j] == '/') {
                concat[j] = '_';
            }
        }
        DumpCategory(regHandle, std::string("/") + topLevel[i], concat, topLevel[i]);
    }
}

extern "C" int main(int argc, char *argv[]) {
	checkpointNext("Registry test");

    struct RegParam reg;
	REGHANDLE regHandle;
	int ret = 0;

	memset(&reg, 0, sizeof(reg));
	reg.regtype = 1;
	reg.namelen = strlen("/system");
	reg.unk2 = 1;
	reg.unk3 = 1;
	strcpy(reg.name, "/syst1m");

    int retval = sceRegOpenRegistry(&reg, 2, &regHandle);
    schedf("%08x = sceRegOpenRegistry(bad) -> handle %08x\n", retval, regHandle);

    // OK, trying with the correct name now.
	strcpy(reg.name, "/system");
    retval = sceRegOpenRegistry(&reg, 2, &regHandle);
    schedf("%08x = sceRegOpenRegistry(/system) -> handle %08x\n", retval, regHandle);
    if (retval < 0) {
        schedf("couldn't open registry: %08X\n", retval);
        // OK, we're screwed.
        return 0;
    }


    REGHANDLE rootCategory;
    retval = sceRegOpenCategory(regHandle, "", 2, &rootCategory);
    schedf("%08x = sceRegOpenCategory() -> handle %08x\n", retval, rootCategory);

    if (retval < 0) {
        retval = sceRegOpenCategory(regHandle, "/", 2, &rootCategory);
        schedf("%08x = sceRegOpenCategory(/) -> handle %08x\n", retval, rootCategory);
    }

    REGHANDLE fontCategory;

    retval = sceRegOpenCategory(-1337, "/DATA/FONT", 2, &fontCategory);
    schedf("%08x = sceRegOpenCategory(badreg, DATA/FONT) -> handle %08x\n", retval, fontCategory);

    retval = sceRegOpenCategory(regHandle, "/DATA/FANT", 2, &fontCategory);
    schedf("%08x = sceRegOpenCategory(DATA/FANT (bad)) -> handle %08x\n", retval, fontCategory);
    // Ignoring retval here, we tried to provoke an error.

    retval = sceRegOpenCategory(regHandle, "/DATA/FONT", 2, &fontCategory);
    schedf("%08x = sceRegOpenCategory(DATA/FONT)\n", retval);
    if (retval < 0) {
        return 0;
    }
    int numKeys;
    int result = sceRegGetKeysNum(fontCategory, &numKeys);
    schedf("%08x = sceRegGetKeysNum(fontCategory) -> %d\n", result, numKeys);

    char *keyData = (char *)malloc(numKeys * 27);  // ??? 27 bytes per key?
    result = sceRegGetKeys(fontCategory, keyData, numKeys);

    for (int i = 0; i < numKeys; i++) {
        char *keyName = keyData + i * 27;
        schedf("keyname %d: %s\n", i, keyName);unsigned int type;
        SceSize size = 0xcccccccc;
        unsigned char data[1024];
        REGHANDLE keyHandle;
        if (!sceRegGetKeyInfo(fontCategory, keyName, &keyHandle, &type, &size)) {
            if (type == REG_TYPE_DIR) {
                schedf("DIR - %-27s size=%08x\n", keyName, size);
                /*
                int retval = sceRegGetKeyValue(fontCategory, keyHandle, data, size);
                if (retval < 0) {
                    schedf("sceRegGetKeyValue on dir failed: %08X\n", retval);
                } else {
                    schedf("DIR - %-27s (%d)\n", keyName, size);
                }*/
            } else {
                if (!sceRegGetKeyValue(fontCategory, keyHandle, data, size)) {
                    switch(type) {
                    case REG_TYPE_INT:
                        schedf("INT - %-27s - %4d : %d\n", keyName, size, *((int*) data));
                        break;
                    case REG_TYPE_STR:
                        schedf("STR - %-27s - %4d : %s\n", keyName, size, (char *)data);
                        break;
                    case REG_TYPE_BIN:
                        {
                            int i;
                            schedf("BIN - %-27s - %4d : ", keyName, size);
                            for (i = 0; i < size-1; i++) {
                                schedf("%02X-", data[i]);
                            }
                            schedf("%02X\n", data[i]);
                        }
                        break;
                    };
                }
            }
        }
    }

    sceRegCloseCategory(fontCategory);

    DumpRegistry(regHandle);

    retval = sceRegCloseRegistry(1337);
    schedf("%08x = sceRegCloseRegistry(1337)\n", retval);
    retval = sceRegCloseRegistry(regHandle);
    schedf("%08x = sceRegCloseRegistry(regHandle)\n", retval);

	return 0;
}