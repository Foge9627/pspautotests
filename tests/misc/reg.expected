[x] Registry test
00000000 = sceRegOpenRegistry(bad) -> handle 00000000
00000000 = sceRegOpenRegistry(/system) -> handle 00000000
80082716 = sceRegOpenCategory() -> handle ffffffff
80082716 = sceRegOpenCategory(/) -> handle ffffffff
80082715 = sceRegOpenCategory(badreg, DATA/FONT) -> handle ffffffff
80082718 = sceRegOpenCategory(DATA/FANT (bad)) -> handle ffffffff
00000000 = sceRegOpenCategory(DATA/FONT)
00000000 = sceRegGetKeysNum(fontCategory) -> 3
keyname 0: path_name
STR - path_name                   -   13 : flash0:/font
keyname 1: num_fonts
INT - num_fonts                   -    4 : 18
keyname 2: PROPERTY
DIR - PROPERTY                    size=00000000


// Dump of the PSP registry using tests/misc/reg.prx in pspautotests

// Dump of /DATA/FONT/PROPERTY/INFO0
static const KeyValue tree_DATA_FONT_PROPERTY_INFO0[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000001 },
	{ "style", ValueType::INT, "", (int)0x00000067 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000001 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-NewRodin Pro DB" },
	{ "file_name", ValueType::STR, "jpn0.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO1
static const KeyValue tree_DATA_FONT_PROPERTY_INFO1[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000001 },
	{ "style", ValueType::INT, "", (int)0x00000001 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-NewRodin Pro Latin" },
	{ "file_name", ValueType::STR, "ltn0.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO2
static const KeyValue tree_DATA_FONT_PROPERTY_INFO2[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000002 },
	{ "style", ValueType::INT, "", (int)0x00000001 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-Matisse Pro Latin" },
	{ "file_name", ValueType::STR, "ltn1.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO3
static const KeyValue tree_DATA_FONT_PROPERTY_INFO3[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000001 },
	{ "style", ValueType::INT, "", (int)0x00000002 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-NewRodin Pro Latin" },
	{ "file_name", ValueType::STR, "ltn2.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO4
static const KeyValue tree_DATA_FONT_PROPERTY_INFO4[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000002 },
	{ "style", ValueType::INT, "", (int)0x00000002 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-Matisse Pro Latin" },
	{ "file_name", ValueType::STR, "ltn3.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO5
static const KeyValue tree_DATA_FONT_PROPERTY_INFO5[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000001 },
	{ "style", ValueType::INT, "", (int)0x00000005 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-NewRodin Pro Latin" },
	{ "file_name", ValueType::STR, "ltn4.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO6
static const KeyValue tree_DATA_FONT_PROPERTY_INFO6[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000002 },
	{ "style", ValueType::INT, "", (int)0x00000005 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-Matisse Pro Latin" },
	{ "file_name", ValueType::STR, "ltn5.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO7
static const KeyValue tree_DATA_FONT_PROPERTY_INFO7[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000001 },
	{ "style", ValueType::INT, "", (int)0x00000006 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-NewRodin Pro Latin" },
	{ "file_name", ValueType::STR, "ltn6.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO8
static const KeyValue tree_DATA_FONT_PROPERTY_INFO8[] = {
	{ "h_size", ValueType::INT, "", (int)0x00000288 },
	{ "v_size", ValueType::INT, "", (int)0x00000288 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000002 },
	{ "style", ValueType::INT, "", (int)0x00000006 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-Matisse Pro Latin" },
	{ "file_name", ValueType::STR, "ltn7.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO9
static const KeyValue tree_DATA_FONT_PROPERTY_INFO9[] = {
	{ "h_size", ValueType::INT, "", (int)0x000001c0 },
	{ "v_size", ValueType::INT, "", (int)0x000001c0 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000001 },
	{ "style", ValueType::INT, "", (int)0x00000001 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-NewRodin Pro Latin" },
	{ "file_name", ValueType::STR, "ltn8.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO10
static const KeyValue tree_DATA_FONT_PROPERTY_INFO10[] = {
	{ "h_size", ValueType::INT, "", (int)0x000001c0 },
	{ "v_size", ValueType::INT, "", (int)0x000001c0 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000002 },
	{ "style", ValueType::INT, "", (int)0x00000001 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-Matisse Pro Latin" },
	{ "file_name", ValueType::STR, "ltn9.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO11
static const KeyValue tree_DATA_FONT_PROPERTY_INFO11[] = {
	{ "h_size", ValueType::INT, "", (int)0x000001c0 },
	{ "v_size", ValueType::INT, "", (int)0x000001c0 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000001 },
	{ "style", ValueType::INT, "", (int)0x00000002 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-NewRodin Pro Latin" },
	{ "file_name", ValueType::STR, "ltn10.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA/FONT/PROPERTY/INFO12
static const KeyValue tree_DATA_FONT_PROPERTY_INFO12[] = {
	{ "h_size", ValueType::INT, "", (int)0x000001c0 },
	{ "v_size", ValueType::INT, "", (int)0x000001c0 },
	{ "h_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "v_resolution", ValueType::INT, "", (int)0x00002000 },
	{ "extra_attributes", ValueType::INT, "", (int)0x00000000 },
	{ "weight", ValueType::INT, "", (int)0x00000000 },
	{ "family_code", ValueType::INT, "", (int)0x00000002 },
	{ "style", ValueType::INT, "", (int)0x00000002 },
	{ "sub_style", ValueType::INT, "", (int)0x00000000 },
	{ "language_code", ValueType::INT, "", (int)0x00000002 },
	{ "region_code", ValueType::INT, "", (int)0x00000000 },
	{ "country_code", ValueType::INT, "", (int)0x00000001 },
	{ "font_name", ValueType::STR, "FTT-Matisse Pro Latin" },
	{ "file_name", ValueType::STR, "ltn11.pgf" },
	{ "expire_date", ValueType::INT, "", (int)0x00000000 },
	{ "shadow_option", ValueType::INT, "", (int)0x00000000 },
};

// /DATA/FONT/PROPERTY/INFO13 was not accessible (returned 80082712)
static const KeyValue tree_DATA_FONT_PROPERTY_INFO13[1] = { "", ValueType::FAIL, "", (int)0x80082712 };

// /DATA/FONT/PROPERTY/INFO14 was not accessible (returned 80082712)
static const KeyValue tree_DATA_FONT_PROPERTY_INFO14[1] = { "", ValueType::FAIL, "", (int)0x80082712 };

// /DATA/FONT/PROPERTY/INFO15 was not accessible (returned 80082712)
static const KeyValue tree_DATA_FONT_PROPERTY_INFO15[1] = { "", ValueType::FAIL, "", (int)0x80082712 };

// /DATA/FONT/PROPERTY/INFO16 was not accessible (returned 80082712)
static const KeyValue tree_DATA_FONT_PROPERTY_INFO16[1] = { "", ValueType::FAIL, "", (int)0x80082712 };

// /DATA/FONT/PROPERTY/INFO17 was not accessible (returned 80082712)
static const KeyValue tree_DATA_FONT_PROPERTY_INFO17[1] = { "", ValueType::FAIL, "", (int)0x80082712 };

// Dump of /DATA/FONT/PROPERTY
static const KeyValue tree_DATA_FONT_PROPERTY[] = {
	{ "INFO0", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO0), tree_DATA_FONT_PROPERTY_INFO0 },
	{ "INFO1", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO1), tree_DATA_FONT_PROPERTY_INFO1 },
	{ "INFO2", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO2), tree_DATA_FONT_PROPERTY_INFO2 },
	{ "INFO3", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO3), tree_DATA_FONT_PROPERTY_INFO3 },
	{ "INFO4", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO4), tree_DATA_FONT_PROPERTY_INFO4 },
	{ "INFO5", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO5), tree_DATA_FONT_PROPERTY_INFO5 },
	{ "INFO6", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO6), tree_DATA_FONT_PROPERTY_INFO6 },
	{ "INFO7", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO7), tree_DATA_FONT_PROPERTY_INFO7 },
	{ "INFO8", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO8), tree_DATA_FONT_PROPERTY_INFO8 },
	{ "INFO9", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO9), tree_DATA_FONT_PROPERTY_INFO9 },
	{ "INFO10", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO10), tree_DATA_FONT_PROPERTY_INFO10 },
	{ "INFO11", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO11), tree_DATA_FONT_PROPERTY_INFO11 },
	{ "INFO12", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO12), tree_DATA_FONT_PROPERTY_INFO12 },
	{ "INFO13", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO13), tree_DATA_FONT_PROPERTY_INFO13 },
	{ "INFO14", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO14), tree_DATA_FONT_PROPERTY_INFO14 },
	{ "INFO15", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO15), tree_DATA_FONT_PROPERTY_INFO15 },
	{ "INFO16", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO16), tree_DATA_FONT_PROPERTY_INFO16 },
	{ "INFO17", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY_INFO17), tree_DATA_FONT_PROPERTY_INFO17 },
};

// Dump of /DATA/FONT
static const KeyValue tree_DATA_FONT[] = {
	{ "path_name", ValueType::STR, "flash0:/font" },
	{ "num_fonts", ValueType::INT, "", (int)0x00000012 },
	{ "PROPERTY", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT_PROPERTY), tree_DATA_FONT_PROPERTY },
};

// Dump of /DATA/COUNT
static const KeyValue tree_DATA_COUNT[] = {
	{ "boot_count", ValueType::INT, "", (int)0x00000000 },
	{ "game_exec_count", ValueType::INT, "", (int)0x00000046 },
	{ "slide_count", ValueType::INT, "", (int)0x00000000 },
	{ "usb_connect_count", ValueType::INT, "", (int)0x000000ec },
	{ "wifi_connect_count", ValueType::INT, "", (int)0x00000000 },
	{ "psn_access_count", ValueType::INT, "", (int)0x00000000 },
};

// Dump of /DATA
static const KeyValue tree_DATA[] = {
	{ "FONT", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_FONT), tree_DATA_FONT },
	{ "COUNT", ValueType::DIR, "", ARRAY_SIZE(tree_DATA_COUNT), tree_DATA_COUNT },
};

// /SYSPROFILE was not accessible (returned 80082712)
static const KeyValue tree_SYSPROFILE[1] = { "", ValueType::FAIL, "", (int)0x80082712 };

80082715 = sceRegCloseRegistry(1337)
00000000 = sceRegCloseRegistry(regHandle)
